File: src/bin/day2_1.rs
use std::fs::File;
use std::io::{BufReader, BufRead};

const ROCK: usize = 1;
const PAPER: usize = 2;
const SIZ: usize = 3;

fn get_score_of_shape(shape: &str)-> usize{
    if shape == "A" || shape == "X" {
	ROCK
    }else if shape == "B" || shape == "Y" {
	PAPER
    }else if shape == "C" || shape == "Z" {
	SIZ
    }else {
	panic!("Should never reach here");
    }
}

fn get_score(them: &str, me: &str)-> usize{
    if me == "X" && them == "A" || me == "Y" && them == "B" || me == "Z" && them == "C" {
	return 3 + get_score_of_shape(them)
    }
    if me == "X" && them == "C"{
	return 6 + get_score_of_shape(me);
    }else if me == "Z" && them == "A"{
	return 0 + get_score_of_shape(me);
    }

    if me == "Z" && them == "B"{
	return 6 + get_score_of_shape(me);
    }else if me == "Y" && them == "C"{
	return 0 + get_score_of_shape(me);
    }

    if me == "Y" && them == "A"{
	return 6 + get_score_of_shape(me);
    }else if me == "X" && them == "B"{
	return 0 + get_score_of_shape(me);
    }
    println!("WTF!!! {} {}",me, them);
    0
}

fn main(){
    let mut score: usize = 0;
    for l in BufReader::new(File::open("./day2_1.txt").unwrap()).lines(){
	let line = l.unwrap();
	let vals = line.split(' ').collect::<Vec<&str>>();
	let (them,me) = (vals[0],vals[1]);
	score+=get_score(them,me);
    }
    println!("Total: {score}");
}

File: src/day2_1.zig
const std = @import("std");
const utils = @import("utils");
const ArrayList = std.ArrayList;

// Rock > Siz
// Siz > Paper
// Paper > Rock

// A = Rock = 1 = X
// B = Paper = 2 = Y
// C = Siz = 3 = Z

// Loss = 0
// Win = 6
// Draw = 3

fn scoreOfChoice(choice: u8) usize{
    switch(choice){
        88 => return 1,
        89 => return 2,
        90 => return 3,
        else => unreachable
    }
}

fn getScore(op: u8, me: u8) usize {
    //win
    if((me-23 == 65  and op==67 ) or 
        (me-23 == 67 and  op == 66) or 
        (me-23 == 66 and op == 65)){
        return 6+scoreOfChoice(me);
    }
    //draw
    else if(op == (me-23)){
        return 3+scoreOfChoice(me);
    }
    return scoreOfChoice(me);
}

pub fn main() !void{
    const stdout = std.io.getStdOut().writer();
    try stdout.print("Day 2 p1 solution\n", .{});
    var lines: ArrayList(ArrayList(u8)) = try utils.readLinesFromFile(100, "res/day2_1.txt");
    defer lines.deinit();

    var score: usize = 0;

    for(lines.items)|line|{
        if(line.items.len==0){
            continue;
        }
        const op = line.items[0];
        const me = line.items[2];
        const result_score = getScore(op,me);
        score += result_score;
    }
    try stdout.print("Total score is {d}\n",.{score});
}
